tests/promise/test_dataloader.py

```diff
@@ -0,0 +1,250 @@
+
+
+def id_loader(**options):
+    load_calls = []
+
+    resolve = options.pop("resolve", Promise.resolve)
+
+    def fn(keys):
+        load_calls.append(keys)
+        return resolve(keys)
+
+    identity_loader = PromiseDataLoader(fn, **options)
+    return identity_loader, load_calls
+
+
+def test_build_a_simple_data_loader():
+    def call_fn(keys):
+        return Promise.resolve(keys)
+
+    identity_loader = PromiseDataLoader(call_fn)
+
+    promise1 = identity_loader.load(1)
+    assert isinstance(promise1, Promise)
+
+    value1 = promise1.get()
+    assert value1 == 1
+
+
+def test_batches_multiple_requests():
+    @Promise.safe
+    def do():
+        identity_loader, load_calls = id_loader()
+
+        promise1 = identity_loader.load(1)
+        promise2 = identity_loader.load(2)
+
+        p = Promise.all([promise1, promise2])
+
+        value1, value2 = p.get()
+
+        assert value1 == 1
+        assert value2 == 2
+
+        assert load_calls == [[1, 2]]
+
+    do().get()
+
+
+def test_batches_multiple_requests_with_max_batch_sizes():
+    @Promise.safe
+    def do():
+        identity_loader, load_calls = id_loader(max_batch_size=2)
+
+        promise1 = identity_loader.load(1)
+        promise2 = identity_loader.load(2)
+        promise3 = identity_loader.load(3)
+
+        p = Promise.all([promise1, promise2, promise3])
+
+        value1, value2, value3 = p.get()
+
+        assert value1 == 1
+        assert value2 == 2
+        assert value3 == 3
+
+        assert load_calls == [[1, 2], [3]]
+
+    do().get()
+
+
+def test_coalesces_identical_requests():
+    @Promise.safe
+    def do():
+        identity_loader, load_calls = id_loader()
+
+        promise1 = identity_loader.load(1)
+        promise2 = identity_loader.load(1)
+
+        assert promise1 == promise2
+        p = Promise.all([promise1, promise2])
+
+        value1, value2 = p.get()
+
+        assert value1 == 1
+        assert value2 == 1
+
+        assert load_calls == [[1]]
+
+    do().get()
+
+
+def test_caches_repeated_requests():
+    @Promise.safe
+    def do():
+        identity_loader, load_calls = id_loader()
+
+        a, b = Promise.all([identity_loader.load("A"), identity_loader.load("B")]).get()
+
+        assert a == "A"
+        assert b == "B"
+
+        assert load_calls == [["A", "B"]]
+
+        a2, c = Promise.all(
+            [identity_loader.load("A"), identity_loader.load("C")]
+        ).get()
+
+        assert a2 == "A"
+        assert c == "C"
+
+        assert load_calls == [["A", "B"], ["C"]]
+
+        a3, b2, c2 = Promise.all(
+            [
+                identity_loader.load("A"),
+                identity_loader.load("B"),
+                identity_loader.load("C"),
+            ]
+        ).get()
+
+        assert a3 == "A"
+        assert b2 == "B"
+        assert c2 == "C"
+
+        assert load_calls == [["A", "B"], ["C"]]
+
+    do().get()
+
+
+# Represents Errors
+
+
+def test_resolves_to_error_to_indicate_failure():
```

Comment:
Was this comment supposed to be attached to the below function, and the autoformatter separated it?

Reply:
Even if so, I'm not too sure what it means